### ISSUE TITLE
ZScheduler (dead)lock

### ISSUE THREAD

**Issue by zeal18** [zio/zio#7930]
Hi all!
We are facing a rare issue but it started to happen periodically (once in a day or two) with one of our services.

### Symptoms
ZIO just stops working, nothing happens, the service just hangs doing nothing. Netty continues accepting requests (it has it's own thread pool) which hang trying to interrupt fibers by timeout. CPU doesn't change, threads amount grows depending on incoming traffic, memory grows because of threads amount and some stuck effects.

### ZScheduler state
- **poolSize**: 1
- **state**: 0
- **workers**: contains 1 worker
- **idle**: null
- **globalQueue**: contains a lot of `FiberRuntime`s (enqueued: 539198, dequeued: 539170)
- **cache**: contains a bunch of workers (enqueued: 77, dequeued: 72)
- **blockingLocations**: `EmptySet`
- **submittedLocations**: contains 17 elements

#### Cached workers

All cached workers have the same state:
- **localQueue**: empty
- **nextRunnable**: null
- **currentRunnable**: null
- **blocking**: true
- **active**: false

#### Pool worker

- **localQueue**: empty
- **nextRunnable**: null
- **currentRunnable**: null
- **blocking**: false
- **active**: false

### The lock
As far as I understand, the lock happens because the single worker from the pool becomes nonactive but it's not marked as **idle**, after that the scheduler can schedule nothing and hangs forever

### Possible way to the lock
Here I'm trying to visualise a concurrent run of the worker and submitting a task to it which could lead to the lock

Legend:
`#0969DA` ZScheduler.Worker.run()
`#DA6909` ZScheduler.submit()

The way:
`#0969DA` `} else {`
`#0969DA` `  idle.offer(self)`
`#0969DA` `}`
`#0969DA` `...`
`#DA6909` `val currentSearching = currentState & 0xffff`
`#DA6909` `if (currentActive != poolSize && currentSearching == 0) {`
`#DA6909` `  var loop = true`
`#DA6909` `  while (loop) {`
`#DA6909` `    val worker = idle.poll(null)`
`#DA6909` `    if (worker eq null) {`
`#DA6909` `      loop = false`
`#DA6909` `    } else {`
`#DA6909` `      state.getAndAdd(0x10001)`
`#0969DA` `while (!active && !isInterrupted) {`
`#DA6909` `      worker.active = true`
`#DA6909` `      LockSupport.unpark(worker)`
`#0969DA` `  LockSupport.park()`
`#DA6909` `      loop = false`
`#DA6909` `    }`
`#DA6909` `  }`
`#DA6909` `}`

**Comment by adamgfraser** [comment]
@zeal18 I don't think this is possible because `unpark` "makes available the permit for the given thread, if it was not already available. If the thread was blocked on park then it will unblock. Otherwise, its next call to park is guaranteed not to block." So in the scenario above `park` would return immediately, `active` would be true and the worker would continue with its run loop.

If you can create a small reproducer definitely happy to take a look!

**Comment by unzer-alezhoev** [comment]
> @zeal18 I don't think this is possible because `unpark` "makes available the permit for the given thread, if it was not already available. If the thread was blocked on park then it will unblock. Otherwise, its next call to park is guaranteed not to block." So in the scenario above `park` would return immediately, `active` would be true and the worker would continue with its run loop.
> 
> If you can create a small reproducer definitely happy to take a look!

totally makes sense, unfortunately such a concurrency issue is barely reproducible because it could depend on the environment it runs (a node, other containers deployed on the same node and load on them, etc). 

According to the data I can see in the heap dump it's parked, it's on the `workers` array, it's not active, but it's not in the `idle` as well. I'll try to find another possible way to this state

**Comment by zeal18** [comment]
> @zeal18 I don't think this is possible because `unpark` "makes available the permit for the given thread, if it was not already available. If the thread was blocked on park then it will unblock. Otherwise, its next call to park is guaranteed not to block." So in the scenario above `park` would return immediately, `active` would be true and the worker would continue with its run loop.
> 
> If you can create a small reproducer definitely happy to take a look!

totally makes sense, unfortunately such a concurrency issue is barely reproducible because it could depend on the environment it runs (a node, other containers deployed on the same node and load on them, etc).

According to the data I can see in the heap dump it's parked, it's on the `workers` array, it's not `active`, but it's not in the `idle` as well. I'll try to find another possible way to this state

**Comment by zeal18** [comment]
Hi @adamgfraser! I've tried to find another way to reach the "stuck" state but had no success. At the same time I could observe the same issue a couple of times on the prod after reporting it here. 
In case I don't have a proof that it's a ZIO bug, I think it makes sense to close this issue

**Comment by adamgfraser** [comment]
@zeal18 Okay. Will be on the lookout as well.

**Comment by ipetkovic** [comment]
I had exactly the same issue.
Putting comment here if someone else stumbles upon it.
It should be solved with:
https://github.com/zio/zio/pull/8552
